aggregator/generic_elem.go

```diff
@@ -0,0 +1,315 @@
+		e.toConsume[i].aggregation.Close()
+		e.toConsume[i].Reset()
+	}
+
+	return canCollect
+}
+
+// Close closes the element.
+func (e *GenericElem) Close() {
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return
+	}
+	e.closed = true
+	e.id = nil
+	for idx := range e.values {
+		// Close the underlying aggregation objects.
+		e.values[idx].aggregation.Close()
+		e.values[idx].Reset()
+	}
+	e.values = e.values[:0]
+	e.toConsume = e.toConsume[:0]
+	e.genericElemBase.Close()
+	aggTypesPool := e.aggTypesOpts.TypesPool()
+	pool := e.ElemPool(e.opts)
+	e.Unlock()
+
+	if !e.useDefaultAggregation {
+		aggTypesPool.Put(e.aggTypes)
+	}
+	pool.Put(e)
+}
+
+// findOrCreate finds the aggregation for a given time, or creates one
+// if it doesn't exist.
+func (e *GenericElem) findOrCreate(alignedStart int64) (lockedAggregation, error) {
+	e.RLock()
+	if e.closed {
+		e.RUnlock()
+		return nil, errElemClosed
+	}
+	idx, found := e.indexOfWithLock(alignedStart)
+	if found {
+		agg := e.values[idx].aggregation
+		e.RUnlock()
+		return agg, nil
+	}
+	e.RUnlock()
+
+	e.Lock()
+	if e.closed {
+		e.Unlock()
+		return nil, errElemClosed
+	}
+	idx, found = e.indexOfWithLock(alignedStart)
+	if found {
+		agg := e.values[idx].aggregation
+		e.Unlock()
+		return agg, nil
+	}
+
+	// If not found, create a new aggregation.
+	numValues := len(e.values)
+	e.values = append(e.values, timedAggregation{})
+	copy(e.values[idx+1:numValues+1], e.values[idx:numValues])
+	e.values[idx] = timedAggregation{
+		timeNanos:   alignedStart,
+		aggregation: e.NewLockedAggregation(e.opts, e.aggOpts),
+	}
+	agg := e.values[idx].aggregation
+	e.Unlock()
+	return agg, nil
+}
+
+// indexOfWithLock finds the smallest element index whose timestamp
+// is no smaller than the start time passed in, and true if it's an
+// exact match, false otherwise.
+func (e *GenericElem) indexOfWithLock(alignedStart int64) (int, bool) {
+	numValues := len(e.values)
+	// Optimize for the common case.
+	if numValues > 0 && e.values[numValues-1].timeNanos == alignedStart {
+		return numValues - 1, true
+	}
+	// Binary search for the unusual case. We intentionally do not
+	// use the sort.Search() function because it requires passing
+	// in a closure.
+	left, right := 0, numValues
+	for left < right {
+		mid := left + (right-left)/2 // avoid overflow
+		if e.values[mid].timeNanos < alignedStart {
+			left = mid + 1
+		} else {
+			right = mid
+		}
+	}
+	// If the current timestamp is equal to or larger than the target time,
+	// return the index as is.
+	if left < numValues && e.values[left].timeNanos == alignedStart {
+		return left, true
+	}
+	return left, false
+}
+
+func (e *GenericElem) processValue(timeNanos int64, agg lockedAggregation, fn aggMetricFn) {
+	var fullPrefix = e.FullPrefix(e.opts)
+	if e.useDefaultAggregation {
+		// NB(cw) Use default suffix slice for faster look up.
+		suffixes := e.DefaultAggregationTypeStrings(e.aggTypesOpts)
+		aggTypes := e.DefaultAggregationTypes(e.aggTypesOpts)
+		agg.Lock()
```

Comment:
did we have this lock before? why do we need it now?

Reply:
Yeah it was there before (in `Consume` in `elem.go`). This is simply moving it into the `processValues` method.